### PR TITLE
UrlHelper - Added standard support for UTF8 encoding, added support for RFC2396  &  RFC3986

### PR DESCRIPTION
**Pull request by snakefoot**
Grapped the unit-test from #1725, but implemented the encoding manually so it stays the same independent of dot-net version.

Fixes https://github.com/NLog/NLog/issues/1724

---

This change is [Reviewable](https://reviewable.io/reviews/nlog/nlog/1735)

### CODE REVIEW

**Comment by 304NotModified** [comment]
Cool! Will check this! Thanks!

File: src/NLog/Internal/UrlHelper.cs [hunk 1]
```diff
@@ -33,82 +33,152 @@
...
 
         /// <summary>
-        /// Url encode and URL
+        /// Url encode an URL
         /// </summary>
         /// <param name="str">URL to be encoded</param>
         /// <param name="spaceAsPlus">space as + or %20? <c>false</c> (%20) is the safe option.</param>
         /// <returns>Encoded url.</returns>
-        internal static string UrlEncode(string str, bool spaceAsPlus)
+        public static string UrlEncode(string str, bool spaceAsPlus)
```
Comment (304NotModified) [inline src/NLog/Internal/UrlHelper.cs]: Why public?

File: src/NLog/Internal/UrlHelper.cs [hunk 2]
```diff
@@ -33,82 +33,152 @@
...
+        /// Escape unicode string data for use in http-requests
+        /// </summary>
+        /// <param name="target">unicode string-data to be encoded</param>
+        /// <param name="flags"><see cref="EscapeEncodingFlag"/>s for how to perform the encoding</param>
+        public static void EscapeDataEncode(StringBuilder target, EscapeEncodingFlag flags)
+        {
+            if (target.Length == 0)
+                return;
+
+            bool isUriString = ((int)flags & (int)EscapeEncodingFlag.UriString) == (int)EscapeEncodingFlag.UriString;
```
Comment (304NotModified) [inline src/NLog/Internal/UrlHelper.cs]: I think it's performance wise to create a variable for `(int)flags`?  *maybe also for others)

File: src/NLog/Internal/UrlHelper.cs (hunk 1 above)
Comment (snakefoot) [inline src/NLog/Internal/UrlHelper.cs]: Well it is still hidden by the internal class protection, and it is meant to be used outside this class

File: src/NLog/Internal/UrlHelper.cs (hunk 2 above)
Comment (snakefoot) [inline src/NLog/Internal/UrlHelper.cs]: Not sure I understand. Is there a performance hit by converting an enum to an integer? Actually thought I avoided the performance hit of the normal HasFlag() method :)

Comment (304NotModified) [inline src/NLog/Internal/UrlHelper.cs]: Casting in general could take some performance. Not sure in this case.

**Comment by snakefoot** [comment]
Have added some properties to UrlEncodeLayoutRendererWrapper and WebServiceTarget, so one can control the encoding/escaping.

By default Rfc2396 is enabled, as this was the standard the existing NLog tried to implement. But I have changed so one have to enable EscapeDataNLogLegacy to get the original NLog encoding.

**Comment by 304NotModified** [comment]
Great!

Do you think we should release this in 4.3.11 or 4.4?

**Comment by snakefoot** [comment]
No idea about the timing. The default config chosen by me, makes it a breaking change.

I guess the default config could be changed to a non-breaking change (EscapeDataNLogLegacy  = default true), depending on the urgency for 4.3.11.

**Comment by 304NotModified** [comment]
What's your opinion on EscapeDataNLogLegacy ? Is that a plausible option or just buggy legacy stuff? If it's a bug, then it isn't a breaking change :)

(Very good you have added the option, don't wanna change that)

**Comment by snakefoot** [comment]
I would say EscapeDataNLogLegacy is non-standard legacy stuff. Normal url-data without unicode characters and "unwise" special characters will not be affected by this change.

**Comment by 304NotModified** [comment]
Reviewed 1 of 4 files at r1, 3 of 4 files at r3.
Review status: 4 of 6 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/nlog/nlog/1735)*

**Comment by 304NotModified** [comment]
Reviewed 1 of 4 files at r1, 1 of 4 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/nlog/nlog/1735)*

**Comment by 304NotModified** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[src/NLog/Internal/UrlHelper.cs, line 101 at r1](https://reviewable.io:443/reviews/nlog/nlog/1735#-KVM_wS390xBJ8-mZWw0-r1-101:-KVv_Ral5vXKmrg_utys:b-dqd85w) ([raw file](https://github.com/nlog/nlog/blob/fe5eab6b5fe12db6e8a4a6723cdcac959e2f3f7e/src/NLog/Internal/UrlHelper.cs#L101)):*
<details><summary><i>Previously, 304NotModified (Julian Verdurmen) wrote…</i></summary>
> Casting in general could take some performance. Not sure in this case. 
</details>
Checked it, in this case no performance hit.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/nlog/nlog/1735)*

**Comment by 304NotModified** [comment]
Will merge this for NLog 4.4! Thanks for this great contribution! 

We need some docs here, right?

**Comment by snakefoot** [comment]
> We need some docs here, right?
Yes breaking change along with new parameters for url-encode-wrapper and WebServiceTarget to revert to legacy mode.

Update documentation now, and say available from ver. 4.4?

**Comment by 304NotModified** [comment]
yes 4.4 and not really a breaking change as IMO it only fixes a bug. 

But because of all the new settings and RFCs, 4.4 (instead of 4.3.x) is IMO a better choice.

File: tests/NLog.UnitTests/Internal/UrlHelperTests.cs [hunk 3]
```diff
@@ -53,15 +51,54 @@ public class UrlHelperTests
...
+        [Theory]
+        [InlineData("", true, "")]
+        [InlineData("", false, "")]
+        [InlineData("ab cd", true, "ab+cd")]
+        [InlineData("ab cd", false, "ab%20cd")]
+        [InlineData("one&two =three", true, "one%26two+%3dthree")]
+        [InlineData("one&two =three", false, "one%26two%20%3dthree")]
+        [InlineData(".()*-_!ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz", true, ".()*-_!ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz")]
+        [InlineData(".()*-_!ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz", false, ".()*-_!ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz")]
+        public void EscapeDataEncodeTestRfc2396(string input, bool spaceAsPlus, string result)
```
Comment (304NotModified) [inline tests/NLog.UnitTests/Internal/UrlHelperTests.cs]: shouldnt we add some unit test cases for special chars here?

File: tests/NLog.UnitTests/Internal/UrlHelperTests.cs [hunk 4]
```diff
@@ -53,15 +51,54 @@ public class UrlHelperTests
...
+        }
+
+        [Theory]
+        [InlineData("", true, "")]
+        [InlineData("", false, "")]
+        [InlineData("ab cd", true, "ab+cd")]
+        [InlineData("ab cd", false, "ab%20cd")]
+        [InlineData(@"http://foo:bar@host:123/path?query#fragment", true, @"http://foo:bar@host:123/path?query#fragment")]
+        [InlineData(@"http://foo:bar@host:123/path?query#fragment", false, @"http://foo:bar@host:123/path?query#fragment")] 
+        public void EscapeDataEncodeTestRfc3986(string input, bool spaceAsPlus, string result)
```
Comment (304NotModified) [inline tests/NLog.UnitTests/Internal/UrlHelperTests.cs]: same about the special chars

**Comment by 304NotModified** [comment]
Great stuff this!

**Comment by 304NotModified** [comment]
Review status: 5 of 6 files reviewed at latest revision, 2 unresolved discussions.

---

*[tests/NLog.UnitTests/Internal/UrlHelperTests.cs, line 77 at r3](https://reviewable.io:443/reviews/nlog/nlog/1735#-KVN9W8Y_oT1p1fMu2ft-r3-77:-KWcn36Tr-PDD0Xmmv2f:b-7m37tt) ([raw file](https://github.com/nlog/nlog/blob/9f2949c227a6a8c28c366266b2c7bad458ec120a/tests/NLog.UnitTests/Internal/UrlHelperTests.cs#L77)):*
<details><summary><i>Previously, 304NotModified (Julian Verdurmen) wrote…</i></summary>
> shouldnt we add some unit test cases for special chars here?
</details>
I was more thinking of É, é, è etc

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/nlog/nlog/1735)*

**Comment by 304NotModified** [comment]
Review status: 5 of 6 files reviewed at latest revision, 3 unresolved discussions.

---

*[src/NLog/Internal/UrlHelper.cs, line 66 at r4](https://reviewable.io:443/reviews/nlog/nlog/1735#-KWcnZS7HOzohVudn1zy:-KWcnZS7HOzohVudn1zz:btdog2b) ([raw file](https://github.com/nlog/nlog/blob/d28aae4cc290046abe480702452d6ebafe03ebd9/src/NLog/Internal/UrlHelper.cs#L66)):*
> ```Smalltalk
>         /// <param name="target">target for the encoded result</param>
>         /// <param name="flags"><see cref="EscapeEncodingFlag"/>s for how to perform the encoding</param>
>         public static void EscapeDataEncode(string source, StringBuilder target, EscapeEncodingFlag flags)
> ```

I think this looks solid, but maybe we have also to provide a decode, at least for the unit tests? 

What do you think?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/nlog/nlog/1735)*

**Comment by snakefoot** [comment]
*[tests/NLog.UnitTests/Internal/UrlHelperTests.cs, line 77 at r3](https://reviewable.io:443/reviews/nlog/nlog/1735#-KVN9W8Y_oT1p1fMu2ft-r3-77:-KWdbIFOcNQqMppTdc8P:b-896fix) ([raw file](https://github.com/nlog/nlog/blob/9f2949c227a6a8c28c366266b2c7bad458ec120a/tests/NLog.UnitTests/Internal/UrlHelperTests.cs#L77)):*
<details><summary><i>Previously, 304NotModified (Julian Verdurmen) wrote…</i></summary>
> I was more thinking of É, é, è etc
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/nlog/nlog/1735)*

**Comment by snakefoot** [comment]
*[tests/NLog.UnitTests/Internal/UrlHelperTests.cs, line 94 at r3](https://reviewable.io:443/reviews/nlog/nlog/1735#-KVN9W8Y_oT1p1fMu2ft-r3-94:-KWdbPSkOd1VcyBkRgfH:b-896fix) ([raw file](https://github.com/nlog/nlog/blob/9f2949c227a6a8c28c366266b2c7bad458ec120a/tests/NLog.UnitTests/Internal/UrlHelperTests.cs#L94)):*
<details><summary><i>Previously, 304NotModified (Julian Verdurmen) wrote…</i></summary>
> same about the special chars
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/nlog/nlog/1735)*

**Comment by snakefoot** [comment]
*[src/NLog/Internal/UrlHelper.cs, line 66 at r4](https://reviewable.io:443/reviews/nlog/nlog/1735#-KWcnZS7HOzohVudn1zy:-KWdbTlqiYhyMw5a7cb8:b-896fix) ([raw file](https://github.com/nlog/nlog/blob/d28aae4cc290046abe480702452d6ebafe03ebd9/src/NLog/Internal/UrlHelper.cs#L66)):*
<details><summary><i>Previously, 304NotModified (Julian Verdurmen) wrote…</i></summary>
> I think this looks solid, but maybe we have also to provide a decode, at least for the unit tests? 
> 
> What do you think?
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/nlog/nlog/1735)*

**Comment by 304NotModified** [comment]
Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/nlog/nlog/1735)*

**Comment by 304NotModified** [comment]
Merged! Thanks!!!

**Comment by 304NotModified** [comment]
docs:
https://github.com/NLog/NLog/wiki/WebService-target
https://github.com/NLog/NLog/wiki/Url-Encode-Layout-Renderer

thanks @snakefoot for the docs!

**Pull request merged**